src/viewport_renderer.rs

```diff
@@ -2,23 +2,132 @@ use std::collections::HashMap;
 use std::convert::TryFrom;
 use std::error;
 use std::fmt;
+use std::io;
 use std::mem;
 
 use nalgebra::base::Matrix4;
 
 use wgpu::winit;
 
+macro_rules! include_shader {
+    ($name:expr) => {{
+        include_bytes!(concat!(env!("OUT_DIR"), "/shaders/", $name))
+    }};
+}
+
 const DEPTH_FORMAT: wgpu::TextureFormat = wgpu::TextureFormat::D32Float;
 
+const SHADER_MATCAP_VERT: &[u8] = include_shader!("matcap.vert.spv");
+const SHADER_MATCAP_FRAG: &[u8] = include_shader!("matcap.frag.spv");
+
+const MATCAP_TEXTURE_BYTES: &[u8] = include_bytes!("../resources/matcap.png");
+
+/// The geometry vertex data as uploaded on the GPU.
+///
+/// Positions and normals are internally `[f32; 4]` with the last
+/// component filled in as 0.0 or 1.0 depending on it being a position
+/// or vector.
+#[repr(C)]
 #[derive(Debug, Clone, Copy, PartialEq)]
 pub struct Vertex {
-    pub position: [f32; 3],
+    // These are defined as [f32; 4] for 2 reasons:
+    //
+    // - point vs vector clarity
+    // - padding: vec3 has the same size as vec4 in std140 layout.
+    //   https://www.khronos.org/opengl/wiki/Interface_Block_(GLSL)
+    //
+    // We might decide to pack more useful data than just 0/1 in the
+    // last component later.
+    pub position: [f32; 4],
+    pub normal: [f32; 4],
```

Comment:
Not sure I understand the question. The 4th byte is expected in the shaders, they ask for `vec4` and it has meaning (more on that later).

I think (but am not 100% sure) that if you changed the shaders to only accept `vec3`, creating `matcap_render_pipeline` would or should fail (I don't thing wgpu does any shader related validation yet), because it specifies it's gonna push down a `Float4`. If we passed down data where `position` and `normal` were just a `[f32; 3]`, the first 4 bytes of the next value would be read when reading the first `vec4` - not good. So we have to match what the render pipeline and shaders say with our `#[repr(C)` struct.

The 4th component of a `vec4` for positional data is "a flag" whether the `vec4` is a point or a vector. Multiplication with a matrix in homogeneous coordinates behaves differently: vectors are not translated by a transformation matrix, only rotated and scaled. (I think)